threadpoolctl.py

```diff
@@ -599,6 +211,520 @@ def get_original_num_threads(self):
+
+        Return None if none matches.
+        """
+        for prefix in filename_prefixes:
+            if library_basename.startswith(prefix):
+                return prefix
+        return None
+
+    @classmethod
+    def _get_libc(cls):
+        """Load the lib-C for unix systems."""
+        libc = cls._system_libraries.get("libc")
+        if libc is None:
+            libc_name = find_library("c")
+            if libc_name is None:  # pragma: no cover
+                return None
+            libc = ctypes.CDLL(libc_name, mode=_RTLD_NOLOAD)
+            cls._system_libraries["libc"] = libc
+        return libc
+
+    @classmethod
+    def _get_windll(cls, dll_name):
+        """Load a windows DLL"""
+        dll = cls._system_libraries.get(dll_name)
+        if dll is None:
+            dll = ctypes.WinDLL("{}.dll".format(dll_name))
+            cls._system_libraries[dll_name] = dll
+        return dll
+
+    @classmethod
+    def _realpath(cls, filepath, cache_limit=10000):
+        """Small caching wrapper around os.path.realpath to limit system calls
+        """
+        rpath = cls._realpaths.get(filepath)
+        if rpath is None:
+            rpath = os.path.realpath(filepath)
+            if len(cls._realpaths) < cache_limit:
+                # If we drop support for Python 2.7, we could use
+                # functools.lru_cache with maxsize=10000 instead.
+                cls._realpaths[filepath] = rpath
+        return rpath
+
+
+class _Module(ABC):
+    """Abstract base class for the modules
+
+    A module is represented by the following information:
+      - "user_api" : user API. Possible values are {USER_APIS}.
+      - "internal_api" : internal API. Possible values are {INTERNAL_APIS}.
+      - "prefix" : prefix of the shared library's name.
+      - "filepath" : path to the loaded module.
+      - "version" : version of the library (if available).
+      - "num_threads" : the current thread limit.
+
+    In addition, each module may contain internal_api specific entries.
+    """
+    def __init__(self, filepath=None, prefix=None):
+        self.filepath = filepath
+        self.prefix = prefix
+        self.dynlib = ctypes.CDLL(filepath, mode=_RTLD_NOLOAD)
+        self.version = self.get_version()
+        self.num_threads = self.get_num_threads()
+        self._get_extra_info()
+
+    @property
+    @abstractmethod
+    def user_api(self):
+        pass  # pragma: no cover
+
+    @property
+    @abstractmethod
+    def internal_api(self):
+        pass  # pragma: no cover
+
+    def __eq__(self, other):
+        return self.todict() == other.todict()
+
+    def todict(self):
+        """Return relevant info wrapped in a dict"""
+        return {"user_api": self.user_api,
+                "internal_api": self.internal_api,
+                "prefix": self.prefix,
+                "filepath": self.filepath,
+                "version": self.version,
+                "num_threads": self.num_threads}
```

Comment:
Since we are now testing a lot based on the private API, I think we should have dedicated, non-regression, unit tests for the `_Module.todict` and `_ThreadPoolInfo.todicts` methods to avoid introducing regressions silently.

Reply:
Shouldn't the `todict` method also return the extra info? This could be done by storing the `extra_keys` and looping on them here.